compiler/src/iree/compiler/Codegen/Dialect/GPU/TargetUtils/KnownTargets.cpp

```diff
@@ -61,6 +62,21 @@ struct WgpDetails {
 struct ChipDetails {
   uint32_t wgpCount;
   std::optional<StringRef> sku;
+  // Aggregate chip-level bandwidth in TB/s
+  std::optional<float> peakMemoryBandwidthTBs;
+  // Per-data-type compute performance (TFLOPs/s)
+  std::optional<llvm::SmallDenseMap<ComputeBitwidths, float>> peakPerfTFLOPs;
+
+  ChipDetails(
+      uint32_t wgp, std::optional<llvm::StringRef> s = std::nullopt,
+      std::optional<float> bw = std::nullopt,
+      std::initializer_list<std::pair<ComputeBitwidths, float>> perf = {})
+      : wgpCount(wgp), sku(s), peakMemoryBandwidthTBs(bw) {
+    if (!std::empty(perf)) {
+      peakPerfTFLOPs = llvm::SmallDenseMap<ComputeBitwidths, float>(
+          perf.begin(), perf.end());
+    }
```

Comment:
you don't need to check if it's empty or not
```suggestion
      : wgpCount(wgp), sku(s), peakMemoryBandwidthTBs(bw), peakPerfTFLOPs(llvm::from_range, perf) {}
```

Reply:
I think I need advise here. Note that `peakPerfTFLOPs` is wrapped in `std::optional`. So if I don't do the empty check, then peakPerfTFLops.has_value() will always be true. So this defeat the purpose of having a field wrapped in optional. (Something I'm trying to keep consistent with other member of the struct, such as std::optional<StringRef> sku to indicate this field can be skipped).

Would you suggest to drop the optional entirely for this peakPerfTFLOPs?

Reply:
what's the difference between nullopt and an empty set?

Reply:
No real difference I assume. More like whether someone new reading this code can immediately tell that this field can be non-existent.

I don't have strong opinions either way on this situation.

Reply:
OK, I'm fine with either, just use `.empty()` and the constructor that take a range instead.

Reply:
Sure, I'll drop the optional and make it clear in comment instead.

BTW according to  https://github.com/llvm/llvm-project/blob/06d2d1e15663967f99c7c6c52351ff8210e8b857/llvm/include/llvm/ADT/DenseMap.h#L966-L989, there isn't a constructor that can take `llvm::from_range_t` yet with SmallDenseMap. So I'll continue to use initializer list to initialize.